File: python/src/billing/infrastructure/persistence/sql/repositories.py
"""SQLAlchemy Core によるリポジトリ実装。"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from sqlalchemy import Connection, and_, select
from sqlalchemy.exc import IntegrityError

from billing.domain.ids import CustomerId, InvoiceId, PlanId, SubscriptionId
from billing.domain.invoice import Invoice
from billing.domain.plan import Plan
from billing.domain.subscription import Subscription, SubscriptionStatus
from billing.infrastructure.persistence.errors import DuplicateEntity
from billing.infrastructure.persistence.sql import mappers
from billing.infrastructure.persistence.sql.schema import (
    invoice_lines,
    invoices,
    plans,
    subscriptions,
)
from billing.infrastructure.persistence.timestamps import to_iso
from billing.infrastructure.persistence.versioning import VersionTracker

#: 更新バッチの対象になりうる状態。解約済みは二度と対象にならない。
_LIVE_STATUSES = (
    str(SubscriptionStatus.ACTIVE),
    str(SubscriptionStatus.TRIALING),
    str(SubscriptionStatus.PAST_DUE),
)


class SqlPlanRepository:
    def __init__(self, connection: Connection) -> None:
        self._conn = connection

    def get(self, plan_id: PlanId) -> Plan | None:
        row = self._conn.execute(select(plans).where(plans.c.id == str(plan_id))).one_or_none()
        return None if row is None else mappers.row_to_plan(row)

    def list_all(self) -> list[Plan]:
        rows = self._conn.execute(select(plans).order_by(plans.c.id)).all()
        return [mappers.row_to_plan(row) for row in rows]

    def add(self, plan: Plan) -> None:
        try:
            self._conn.execute(plans.insert().values(**mappers.plan_to_row(plan)))
        except IntegrityError as exc:
            raise DuplicateEntity(f"plan {plan.id!r} already exists") from exc


class SqlSubscriptionRepository:
    def __init__(self, connection: Connection) -> None:
        self._conn = connection
        self._versions = VersionTracker("subscription")

    def get(self, subscription_id: SubscriptionId) -> Subscription | None:
        row = self._conn.execute(
            select(subscriptions).where(subscriptions.c.id == str(subscription_id))
        ).one_or_none()
        if row is None:
            return None
        self._versions.remember(row.id, row.version)
        return mappers.row_to_subscription(row)

    def add(self, subscription: Subscription) -> None:
        try:
            self._conn.execute(
                subscriptions.insert().values(
                    **mappers.subscription_to_row(subscription), version=1
                )
            )
        except IntegrityError as exc:
            # SQLAlchemy 固有の例外をここで止める。上の層に IntegrityError が
            # 漏れると、ユースケースが SQLAlchemy を import する羽目になる。
            raise DuplicateEntity(f"subscription {subscription.id!r} already exists") from exc
        self._versions.remember(str(subscription.id), 1)

    def save(self, subscription: Subscription) -> None:
        entity_id = str(subscription.id)
        expected = self._versions.expected(entity_id)
        result = self._conn.execute(
            subscriptions.update()
            .where(
                and_(
                    subscriptions.c.id == entity_id,
                    subscriptions.c.version == expected,
                )
            )
            .values(**mappers.subscription_to_row(subscription), version=expected + 1)
        )
        if result.rowcount != 1:
            # 誰かが先に更新した。ここで黙って上書きすると、相手の変更が消える。
            raise self._versions.conflict(entity_id)
        self._versions.bump(entity_id)

    def list_due(self, at: datetime, *, limit: int = 100) -> list[Subscription]:
        rows = self._conn.execute(
            select(subscriptions)
            .where(
                and_(
                    subscriptions.c.status.in_(_LIVE_STATUSES),
                    subscriptions.c.period_end <= to_iso(at),
                )
            )
            .order_by(subscriptions.c.period_end)
            .limit(limit)
        ).all()
        for row in rows:
            self._versions.remember(row.id, row.version)
        return [mappers.row_to_subscription(row) for row in rows]

    def list_past_due(self, *, limit: int = 100) -> list[Subscription]:
        rows = self._conn.execute(
            select(subscriptions)
            .where(subscriptions.c.status == str(SubscriptionStatus.PAST_DUE))
            .order_by(subscriptions.c.past_due_since)
            .limit(limit)
        ).all()
        for row in rows:
            self._versions.remember(row.id, row.version)
        return [mappers.row_to_subscription(row) for row in rows]


class SqlInvoiceRepository:
    def __init__(self, connection: Connection) -> None:
        self._conn = connection
        self._versions = VersionTracker("invoice")

    def get(self, invoice_id: InvoiceId) -> Invoice | None:
        row = self._conn.execute(
            select(invoices).where(invoices.c.id == str(invoice_id))
        ).one_or_none()
        if row is None:
            return None
        self._versions.remember(row.id, row.version)
        return self._hydrate(row)

    def add(self, invoice: Invoice) -> None:
        try:
            self._conn.execute(
                invoices.insert().values(**mappers.invoice_to_row(invoice), version=1)
            )
        except IntegrityError as exc:
            # idempotency_key の UNIQUE 制約に当たった場合もここに来る。
            raise DuplicateEntity(
                f"invoice {invoice.id!r} or idempotency key "
                f"{invoice.idempotency_key!r} already exists"
            ) from exc
        line_rows = mappers.invoice_line_rows(invoice)
        if line_rows:
            self._conn.execute(invoice_lines.insert(), line_rows)
        self._versions.remember(str(invoice.id), 1)

    def save(self, invoice: Invoice) -> None:
        # 明細は発行時に確定し、以後変化しない（Invoice に明細を足すメソッドはない）。
        # だから save では見出し行だけを更新する。もし明細が可変になったら、ここは
        # 全削除して入れ直す形に変える必要がある。
        entity_id = str(invoice.id)
        expected = self._versions.expected(entity_id)
        result = self._conn.execute(
            invoices.update()
            .where(and_(invoices.c.id == entity_id, invoices.c.version == expected))
            .values(**mappers.invoice_to_row(invoice), version=expected + 1)
        )
        if result.rowcount != 1:
            raise self._versions.conflict(entity_id)
        self._versions.bump(entity_id)

    def find_by_idempotency_key(self, key: str) -> Invoice | None:
        row = self._conn.execute(
            select(invoices).where(invoices.c.idempotency_key == key)
        ).one_or_none()
        if row is None:
            return None
        self._versions.remember(row.id, row.version)
        return self._hydrate(row)

    def list_for_customer(self, customer_id: CustomerId) -> list[Invoice]:
        rows = self._conn.execute(
            select(invoices)
            .where(invoices.c.customer_id == str(customer_id))
            .order_by(invoices.c.issued_at)
        ).all()
        for row in rows:
            self._versions.remember(row.id, row.version)
        return [self._hydrate(row) for row in rows]

    def _hydrate(self, row: Any) -> Invoice:
        invoice_id = row.id
        line_rows = self._conn.execute(
            select(invoice_lines)
            .where(invoice_lines.c.invoice_id == invoice_id)
            .order_by(invoice_lines.c.position)
        ).all()
        return mappers.row_to_invoice(row, list(line_rows))

File: go/internal/infra/sqlite/repositories.go
package sqlite

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"time"

	"github.com/Yutarotaro/clean-architecture-billing/go/internal/domain"
	"github.com/Yutarotaro/clean-architecture-billing/go/internal/infra/storage"
	"github.com/Yutarotaro/clean-architecture-billing/go/internal/usecase"
	sqlitedriver "modernc.org/sqlite"
)

// liveStatuses は更新バッチの対象になりうる状態。解約済みは二度と対象にならない。
var liveStatuses = []any{
	string(domain.StatusActive),
	string(domain.StatusTrialing),
	string(domain.StatusPastDue),
}

// versionTracker は読み出した集約のバージョンを覚えておく。memory 実装と同じ考え方。
type versionTracker struct {
	entity   string
	versions map[string]int
}

func newVersionTracker(entity string) *versionTracker {
	return &versionTracker{entity: entity, versions: map[string]int{}}
}

func (t *versionTracker) remember(id string, version int) { t.versions[id] = version }

func (t *versionTracker) expected(id string) (int, error) {
	version, ok := t.versions[id]
	if !ok {
		return 0, usecase.ConcurrencyConflict(t.entity, id)
	}
	return version, nil
}

func (t *versionTracker) bump(id string) { t.versions[id]++ }

func (t *versionTracker) conflict(id string) error {
	return usecase.ConcurrencyConflict(t.entity, id)
}

// SQLite の拡張エラーコード。
const (
	// sqliteConstraint は主キーや UNIQUE の衝突（SQLITE_CONSTRAINT）。
	sqliteConstraint = 19
	// sqliteBusySnapshot は SQLITE_BUSY_SNAPSHOT。
	//
	// WAL モードで、古いスナップショットを読んだままのトランザクションが書き込もうと
	// したときに返る。SQLite 自身が lost update を防いでいる状態であり、意味としては
	// 楽観ロックの衝突とまったく同じなので、そのように翻訳する。
	sqliteBusySnapshot = 517
)

// asConflict は SQLite の競合を usecase.ErrConcurrencyConflict に翻訳する。
//
// version 列の突き合わせで検出するより先に、データベースが弾いてくれることがある。
// どちらの経路で分かっても、上の層に届く言葉は同じでなければならない。
func asConflict(err error, entity, id string) error {
	if err == nil {
		return nil
	}
	var sqliteErr *sqlitedriver.Error
	if errors.As(err, &sqliteErr) && sqliteErr.Code() == sqliteBusySnapshot {
		return usecase.ConcurrencyConflict(entity, id)
	}
	return err
}

// asDuplicate は一意制約違反を storage.ErrDuplicate に翻訳する。
//
// ドライバ固有の例外をここで止める。上の層に modernc.org/sqlite の型が漏れると、
// ユースケースがドライバを import する羽目になる。
func asDuplicate(err error, format string, args ...any) error {
	if err == nil {
		return nil
	}
	var sqliteErr *sqlitedriver.Error
	if errors.As(err, &sqliteErr) {
		// 下位 8 ビットが SQLITE_CONSTRAINT(19) なら、主キーか UNIQUE の衝突。
		if sqliteErr.Code()&0xff == sqliteConstraint {
			return fmt.Errorf("%w: %s", storage.ErrDuplicate, fmt.Sprintf(format, args...))
		}
	}
	return err
}

type planRepo struct{ tx *sql.Tx }

func (r *planRepo) Get(ctx context.Context, id domain.PlanID) (*domain.Plan, error) {
	row := r.tx.QueryRowContext(ctx,
		`SELECT id, name, price_amount, price_currency, interval FROM plans WHERE id = ?`,
		string(id))
	plan, err := scanPlan(row)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	return plan, err
}

func (r *planRepo) ListAll(ctx context.Context) ([]domain.Plan, error) {
	rows, err := r.tx.QueryContext(ctx,
		`SELECT id, name, price_amount, price_currency, interval FROM plans ORDER BY id`)
	if err != nil {
		return nil, err
	}
	defer func() { _ = rows.Close() }()

	var plans []domain.Plan
	for rows.Next() {
		plan, err := scanPlan(rows)
		if err != nil {
			return nil, err
		}
		plans = append(plans, *plan)
	}
	return plans, rows.Err()
}

func (r *planRepo) Add(ctx context.Context, plan domain.Plan) error {
	_, err := r.tx.ExecContext(ctx,
		`INSERT INTO plans (id, name, price_amount, price_currency, interval) VALUES (?, ?, ?, ?, ?)`,
		string(plan.ID), plan.Name, plan.Price.Amount, plan.Price.Currency, string(plan.Interval))
	return asDuplicate(err, "plan %q already exists", plan.ID)
}

type scanner interface {
	Scan(dest ...any) error
}

func scanPlan(row scanner) (*domain.Plan, error) {
	var (
		id, name, currency, interval string
		amount                       int64
	)
	if err := row.Scan(&id, &name, &amount, &currency, &interval); err != nil {
		return nil, err
	}
	money, err := domain.NewMoney(amount, currency)
	if err != nil {
		return nil, err
	}
	return &domain.Plan{
		ID:       domain.PlanID(id),
		Name:     name,
		Price:    money,
		Interval: domain.BillingInterval(interval),
	}, nil
}

const subscriptionColumns = `id, customer_id, plan_id, status, period_start, period_end,
	cancel_at_period_end, past_due_since, canceled_at, trial_end, version`

type subscriptionRepo struct {
	tx       *sql.Tx
	versions *versionTracker
}

func (r *subscriptionRepo) Get(
	ctx context.Context, id domain.SubscriptionID,
) (*domain.Subscription, error) {
	row := r.tx.QueryRowContext(ctx,
		`SELECT `+subscriptionColumns+` FROM subscriptions WHERE id = ?`, string(id))
	subscription, err := r.scan(row)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	return subscription, err
}

func (r *subscriptionRepo) Add(ctx context.Context, s *domain.Subscription) error {
	args := append(subscriptionArgs(s), 1)
	_, err := r.tx.ExecContext(ctx,
		`INSERT INTO subscriptions (`+subscriptionColumns+`)
		 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`, args...)
	if err != nil {
		return asDuplicate(err, "subscription %q already exists", s.ID)
	}
	r.versions.remember(string(s.ID), 1)
	return nil
}

func (r *subscriptionRepo) Save(ctx context.Context, s *domain.Subscription) error {
	id := string(s.ID)
	expected, err := r.versions.expected(id)
	if err != nil {
		return err
	}
	args := append(subscriptionArgs(s), expected+1, id, expected)
	result, err := r.tx.ExecContext(ctx,
		`UPDATE subscriptions SET
			id = ?, customer_id = ?, plan_id = ?, status = ?, period_start = ?, period_end = ?,
			cancel_at_period_end = ?, past_due_since = ?, canceled_at = ?, trial_end = ?, version = ?
		 WHERE id = ? AND version = ?`, args...)
	if err != nil {
		return asConflict(err, "subscription", id)
	}
	affected, err := result.RowsAffected()
	if err != nil {
		return err
	}
	if affected != 1 {
		// 誰かが先に更新した。ここで黙って上書きすると、相手の変更が消える。
		return r.versions.conflict(id)
	}
	r.versions.bump(id)
	return nil
}

func (r *subscriptionRepo) ListDue(
	ctx context.Context, at time.Time, limit int,
) ([]*domain.Subscription, error) {
	args := append(append([]any{}, liveStatuses...), storage.FormatTime(at), limit)
	rows, err := r.tx.QueryContext(ctx,
		`SELECT `+subscriptionColumns+` FROM subscriptions
		 WHERE status IN (?, ?, ?) AND period_end <= ?
		 ORDER BY period_end LIMIT ?`, args...)
	if err != nil {
		return nil, err
	}
	return r.collect(rows)
}

func (r *subscriptionRepo) ListPastDue(
	ctx context.Context, limit int,
) ([]*domain.Subscription, error) {
	rows, err := r.tx.QueryContext(ctx,
		`SELECT `+subscriptionColumns+` FROM subscriptions
		 WHERE status = ? ORDER BY past_due_since LIMIT ?`,
		string(domain.StatusPastDue), limit)
	if err != nil {
		return nil, err
	}
	return r.collect(rows)
}

func (r *subscriptionRepo) collect(rows *sql.Rows) ([]*domain.Subscription, error) {
	defer func() { _ = rows.Close() }()
	var found []*domain.Subscription
	for rows.Next() {
		subscription, err := r.scan(rows)
		if err != nil {
			return nil, err
		}
		found = append(found, subscription)
	}
	return found, rows.Err()
}

func (r *subscriptionRepo) scan(row scanner) (*domain.Subscription, error) {
	var raw subscriptionRow
	if err := row.Scan(
		&raw.id, &raw.customerID, &raw.planID, &raw.status, &raw.periodStart, &raw.periodEnd,
		&raw.cancelAtPeriodEnd, &raw.pastDueSince, &raw.canceledAt, &raw.trialEnd, &raw.version,
	); err != nil {
		return nil, err
	}
	r.versions.remember(raw.id, raw.version)
	return raw.toDomain()
}
